File: app/src/main/java/com/example/anchornotes/ui/FilterDialogFragment.java
package com.example.anchornotes.ui;

import android.app.DatePickerDialog;
import android.app.Dialog;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.DialogFragment;

import com.example.anchornotes.R;
import com.example.anchornotes.data.NoteSearchFilter;
import com.example.anchornotes.data.ServiceLocator;
import com.example.anchornotes.data.db.TagDao;
import com.example.anchornotes.data.db.TagEntity;
import com.example.anchornotes.databinding.DialogFilterBinding;

import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.List;
import java.util.Locale;

public class FilterDialogFragment extends DialogFragment {
    public static final String TAG = "FilterDialogFragment";
    public static final String RESULT_KEY = "filter_result";
    public static final String FILTER_EXTRA = "filter";

    private DialogFilterBinding b;
    private final java.util.Set<Long> selectedTagIds = new java.util.HashSet<>();
    private Long fromDate = null, toDate = null;

    public static FilterDialogFragment newInstance() { return new FilterDialogFragment(); }

    @NonNull @Override public Dialog onCreateDialog(@Nullable Bundle savedInstanceState) {
        Dialog d = super.onCreateDialog(savedInstanceState);
        d.setTitle(R.string.filter_notes);
        return d;
    }

    @Nullable @Override public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        b = DialogFilterBinding.inflate(inflater, container, false);
        return b.getRoot();
    }

    @Override public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        // Load tags (synchronously, allowed because DB allows main thread queries)
        TagDao tagDao = ServiceLocator.tagDao(requireContext());
        List<TagEntity> tags = tagDao.getAll();
        populateTags(tags);

        b.btnFromDate.setOnClickListener(v -> showDatePicker(true));
        b.btnToDate.setOnClickListener(v -> showDatePicker(false));

        b.btnApply.setOnClickListener(v -> {
            NoteSearchFilter f = new NoteSearchFilter();
            f.tagIds.addAll(selectedTagIds);
            f.fromDate = fromDate;
            f.toDate = toDate;
            f.hasPhoto = b.checkboxHasPhoto.isChecked() ? Boolean.TRUE : null;
            f.hasVoice = b.checkboxHasVoice.isChecked() ? Boolean.TRUE : null;
            f.hasLocation = b.checkboxHasLocation.isChecked() ? Boolean.TRUE : null;

            Bundle result = new Bundle();
            result.putParcelable(FILTER_EXTRA, f);
            getParentFragmentManager().setFragmentResult(RESULT_KEY, result);
            dismiss();
        });

        b.btnClear.setOnClickListener(v -> {
            selectedTagIds.clear();
            fromDate = toDate = null;
            b.btnFromDate.setText(R.string.select_from_date);
            b.btnToDate.setText(R.string.select_to_date);
            b.checkboxHasPhoto.setChecked(false);
            b.checkboxHasVoice.setChecked(false);
            b.checkboxHasLocation.setChecked(false);
            clearTagChecks();
        });

        b.btnCancel.setOnClickListener(v -> dismiss());
    }

    private void populateTags(List<TagEntity> tags) {
        b.layoutTags.removeAllViews();
        if (tags == null || tags.isEmpty()) {
            TextView tv = new TextView(requireContext());
            tv.setText(getString(R.string.no_tags_available));
            tv.setPadding(16, 16, 16, 16);
            b.layoutTags.addView(tv);
            return;
        }
        for (TagEntity t : tags) {
            CheckBox cb = new CheckBox(requireContext());
            cb.setText(t.name);
            cb.setOnCheckedChangeListener((buttonView, isChecked) -> {
                if (isChecked) selectedTagIds.add(t.id);
                else selectedTagIds.remove(t.id);
            });
            LinearLayout.LayoutParams lp = new LinearLayout.LayoutParams(
                    LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT);
            lp.setMargins(0, 8, 0, 8);
            b.layoutTags.addView(cb, lp);
        }
    }

    private void clearTagChecks() {
        for (int i = 0; i < b.layoutTags.getChildCount(); i++) {
            View v = b.layoutTags.getChildAt(i);
            if (v instanceof CheckBox) ((CheckBox) v).setChecked(false);
        }
    }

    private void showDatePicker(boolean isFrom) {
        Calendar c = Calendar.getInstance();
        new DatePickerDialog(requireContext(), (view, year, month, dayOfMonth) -> {
            Calendar sel = Calendar.getInstance();
            sel.set(year, month, dayOfMonth, 0, 0, 0);
            long ts = sel.getTimeInMillis();
            SimpleDateFormat fmt = new SimpleDateFormat("MM/dd/yyyy", Locale.getDefault());
            if (isFrom) { fromDate = ts; b.btnFromDate.setText(fmt.format(sel.getTime())); }
            else { toDate = ts; b.btnToDate.setText(fmt.format(sel.getTime())); }
        }, c.get(Calendar.YEAR), c.get(Calendar.MONTH), c.get(Calendar.DAY_OF_MONTH)).show();
    }

    @Override public void onDestroyView() { super.onDestroyView(); b = null; }
}

File: app/src/main/java/com/example/anchornotes/ui/FilterDialogFragment.kt
package com.example.anchornotes.ui

import android.app.DatePickerDialog
import android.app.Dialog
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.CheckBox
import android.widget.LinearLayout
import androidx.fragment.app.DialogFragment
import androidx.lifecycle.ViewModelProvider
import com.example.anchornotes.R
import com.example.anchornotes.data.NoteSearchFilter
import com.example.anchornotes.data.db.TagEntity
import com.example.anchornotes.databinding.DialogFilterBinding
import com.example.anchornotes.viewmodel.NoteViewModel
import java.util.Calendar

/**
 * Dialog fragment for filtering notes.
 * Allows selection of tags, date range, and flags (photo, voice, location).
 */
class FilterDialogFragment : DialogFragment() {
    
    private var _binding: DialogFilterBinding? = null
    private val binding get() = _binding!!
    
    private lateinit var viewModel: NoteViewModel
    private val selectedTagIds = mutableSetOf<Long>()
    private var fromDate: Long? = null
    private var toDate: Long? = null
    
    companion object {
        const val TAG = "FilterDialogFragment"
        const val RESULT_KEY = "filter_result"
        const val FILTER_EXTRA = "filter"
        
        fun newInstance(): FilterDialogFragment {
            return FilterDialogFragment()
        }
    }
    
    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val dialog = super.onCreateDialog(savedInstanceState)
        dialog.setTitle(R.string.filter_notes)
        return dialog
    }
    
    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = DialogFilterBinding.inflate(inflater, container, false)
        return binding.root
    }
    
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        
        viewModel = ViewModelProvider(requireActivity())[NoteViewModel::class.java]
        
        // Load tags and populate checkboxes
        viewModel.allTags.observe(viewLifecycleOwner) { tags ->
            populateTags(tags)
        }
        
        // Set up date pickers
        binding.btnFromDate.setOnClickListener {
            showDatePicker(true)
        }
        
        binding.btnToDate.setOnClickListener {
            showDatePicker(false)
        }
        
        // Set up Apply button
        binding.btnApply.setOnClickListener {
            applyFilters()
        }
        
        // Set up Clear button
        binding.btnClear.setOnClickListener {
            clearFilters()
        }
        
        // Set up Cancel button
        binding.btnCancel.setOnClickListener {
            dismiss()
        }
    }
    
    private fun populateTags(tags: List<TagEntity>) {
        binding.layoutTags.removeAllViews()
        
        tags.forEach { tag ->
            val checkBox = CheckBox(requireContext()).apply {
                text = tag.name
                id = View.generateViewId()
                setOnCheckedChangeListener { _, isChecked ->
                    if (isChecked) {
                        selectedTagIds.add(tag.id)
                    } else {
                        selectedTagIds.remove(tag.id)
                    }
                }
            }
            
            val layoutParams = LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                LinearLayout.LayoutParams.WRAP_CONTENT
            ).apply {
                setMargins(0, 8, 0, 8)
            }
            
            binding.layoutTags.addView(checkBox, layoutParams)
        }
        
        if (tags.isEmpty()) {
            binding.layoutTags.addView(android.widget.TextView(requireContext()).apply {
                text = getString(R.string.no_tags_available)
                setPadding(16, 16, 16, 16)
            })
        }
    }
    
    private fun showDatePicker(isFromDate: Boolean) {
        val calendar = Calendar.getInstance()
        val year = calendar.get(Calendar.YEAR)
        val month = calendar.get(Calendar.MONTH)
        val day = calendar.get(Calendar.DAY_OF_MONTH)
        
        DatePickerDialog(
            requireContext(),
            { _, selectedYear, selectedMonth, selectedDay ->
                val cal = Calendar.getInstance().apply {
                    set(selectedYear, selectedMonth, selectedDay, 0, 0, 0)
                    set(Calendar.MILLISECOND, 0)
                }
                val timestamp = cal.timeInMillis
                
                if (isFromDate) {
                    fromDate = timestamp
                    binding.btnFromDate.text = java.text.SimpleDateFormat("MM/dd/yyyy", java.util.Locale.getDefault())
                        .format(cal.time)
                } else {
                    toDate = timestamp
                    binding.btnToDate.text = java.text.SimpleDateFormat("MM/dd/yyyy", java.util.Locale.getDefault())
                        .format(cal.time)
                }
            },
            year,
            month,
            day
        ).show()
    }
    
    private fun applyFilters() {
        val filter = NoteSearchFilter(
            query = null, // Query is managed by HomeFragment
            tagIds = selectedTagIds.toList(),
            fromDate = fromDate,
            toDate = toDate,
            hasPhoto = if (binding.checkboxHasPhoto.isChecked) true else null,
            hasVoice = if (binding.checkboxHasVoice.isChecked) true else null,
            hasLocation = if (binding.checkboxHasLocation.isChecked) true else null
        )
        
        // Send result back to HomeFragment
        val result = Bundle().apply {
            putParcelable(FILTER_EXTRA, filter)
        }
        // Use parentFragmentManager to send result to parent fragment
        parentFragmentManager.setFragmentResult(RESULT_KEY, result)
        
        dismiss()
    }
    
    private fun clearFilters() {
        selectedTagIds.clear()
        fromDate = null
        toDate = null
        binding.btnFromDate.text = getString(R.string.select_from_date)
        binding.btnToDate.text = getString(R.string.select_to_date)
        binding.checkboxHasPhoto.isChecked = false
        binding.checkboxHasVoice.isChecked = false
        binding.checkboxHasLocation.isChecked = false
        
        // Clear tag checkboxes
        for (i in 0 until binding.layoutTags.childCount) {
            val view = binding.layoutTags.getChildAt(i)
            if (view is CheckBox) {
                view.isChecked = false
            }
        }
    }
    
    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}

